### ISSUE TITLE
Modifying the textColor property of a UITextView causes its rx.text and rx.textInput.text properties to immediately emit their current values

### ISSUE THREAD

**Issue by jbeachwood** [ReactiveX/RxSwift#962]
**Short description of the issue**:

Modifying a UITextViews "textColor" property causes its rx.text and rx.textInput.text properties to emit their current values.

**Expected outcome**:

Modifying the textColor property should have no effect besides changing the color of the text.

**What actually happens**:

Modifying the textColor property causes textView.rx.text and textView.rx.textInput.text to emit their current value.

**Self contained code example that reproduces the issue**:
The following code snippet creates an endless loop (textColor is modified, rx.text emits a value, textColor is modified etc.) and continuously prints "Optional("Initial String")". This is obviously a contrived example to demonstrate this issue, and it can be avoided by using distinctUntilChanged(), but this behavior shouldn't be present regardless.

```swift
import UIKit
import RxSwift
import RxCocoa

class ViewController: UIViewController {
    //Bind this to a textView in the xib/storyboard
    @IBOutlet private weak var textView: UITextView!
    
    var rx_bag = DisposeBag()

    override func viewDidLoad() {
        super.viewDidLoad()
        self.bindAndBugOut()
    }
    
    private func bindAndBugOut() {
        self.textView.text = "Initial String"
        self.textView.rx.text.asObservable().subscribe(onNext: { (value) in
            self.textView.textColor = UIColor.black
        }, onError: nil, onCompleted: nil, onDisposed: nil).addDisposableTo(self.rx_bag)
        
        self.textView.rx.text.asObservable().subscribe(onNext: { (value) in
            print(value)
        }, onError: nil, onCompleted: nil, onDisposed: nil).addDisposableTo(self.rx_bag)
        self.textView.textColor = UIColor.blue
    }
}
```

**RxSwift/RxCocoa/RxBlocking/RxTest version/commit**

 RxSwift/RxCocoa/RxBlocking 3.0.0

**Platform/Environment**

  - [X] iOS
  - [ ] macOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] playgrounds

**How easy is to reproduce? (chances of successful reproduce after running the self contained code)**

  - [X] easy, 100% repro
  - [ ] sometimes, 10%-100%
  - [ ] hard, 2% - 10%
  - [ ] extremely hard, %0 - 2%

**Xcode version**:

```
  Xcode 8.1 GM
```

**Installation method**:
  - [X] CocoaPods
  - [ ] Carthage
  - [ ] Git submodules

**I have multiple versions of Xcode installed**:
  (so we can know if this is a potential cause of your issue)
  - [X] yes (Xcode 8.0, Xcode 8.1 GM)
  - [ ] no

**Level of RxSwift knowledge**:
  (this is so we can understand your level of knowledge
    and formulate the response in an appropriate manner)
  - [ ] just starting
  - [ ] I have a small code base
  - [X] I have a significant code base

**Comment by kzaher** [comment]
Hi @jbeachwood ,

I've investigated this, but I'm not sure it's possible.

This is the implementation of `text` property.

```
let source: Observable<String?> = Observable.deferred { [weak textView = self.base] in
            let text = textView?.text
            
            let textChanged = textView?.textStorage
                // This project uses text storage notifications because
                // that's the only way to catch autocorrect changes
                // in all cases. Other suggestions are welcome.
                .rx.didProcessEditingRangeChangeInLength
                //.filter { $0.editedMask == .editedCharacters } <-- can't uncomment this because then IME doesn't work (e.g. chinese)
                // This observe on is here because text storage
                // will emit event while process is not completely done,
                // so rebinding a value will cause an exception to be thrown.
                .observeOn(MainScheduler.asyncInstance)
                .map { _ in
                    return textView?.textStorage.string
                }
                ?? Observable.empty()
            
            return textChanged
                .startWith(text)
        }
```

**Comment by jbeachwood** [comment]
@kzaher 

I agree that it shouldn't be possible, but it clearly is because I am seeing this exact behavior in both my current project and in two new projects I have created. Did you actually try the code snippet I provided? I've uploaded a full project demonstrating this behavior at https://github.com/jbeachwood/RxSwiftTextColorDemonstration. I'm currently on Xcode 8.1 GM but I suspect this is happening on both Xcode 8.0 and Xcode 8.1.

**Comment by jbeachwood** [comment]
@kzaher 

This is caused by your reliance on the NSTextStorageDelegate. [UITextView _setTextColor] causes NSTextStorageDelegate.textStorage(_:didProcessEditing:range:changeInLength:)) to be called because it is modifying the attributes of the txt.

The following code snippet will recreate this, once again with an associated UITextView in the storyboard:
`import UIKit
import RxSwift
import RxCocoa

class ViewController: UIViewController, NSTextStorageDelegate {
    @IBOutlet private weak var textView: UITextView!
    
    var rx_bag = DisposeBag()

    override func viewDidLoad() {
        super.viewDidLoad()
        self.textView.textStorage.delegate = self
        self.textView.textColor = .blue
    }
    
    func textStorage(_ textStorage: NSTextStorage, didProcessEditing editedMask: NSTextStorageEditActions, range editedRange: NSRange, changeInLength delta: Int) {
        print("didProcessEditing Called")
    }
}`

Here is a stack-trace from hitting a breakpoint inside NSTextStorageDelegate.textStorage(_:didProcessEditing:range:changeInLength:)), showing that processEditing is called as a direct result of [UITextView _setTextColor]
`#0	0x0000000104d6c4cb in ViewController.textStorage(NSTextStorage, didProcessEditing : NSTextStorageEditActions, range : _NSRange, changeInLength : Int) -> () at /Users/jackson.beachwood/Documents/Development/Source/Projects/Issue Demonstrations/RxTextViewTextColorBug/RxTextViewTextColorBug/ViewController.swift:25
#1	0x0000000104d6c5da in @objc ViewController.textStorage(NSTextStorage, didProcessEditing : NSTextStorageEditActions, range : _NSRange, changeInLength : Int) -> () ()
#2	0x000000010e341e85 in -[NSTextStorage processEditing] ()
#3	0x000000010e341b01 in -[NSTextStorage endEditing] ()
#4	0x000000010688dc3a in -[UITextView _setTextColor:] ()
#5	0x0000000104d6c385 in ViewController.viewDidLoad() -> () at /Users/jackson.beachwood/Documents/Development/Source/Projects/Issue Demonstrations/RxTextViewTextColorBug/RxTextViewTextColorBug/ViewController.swift:21
#6	0x0000000104d6c482 in @objc ViewController.viewDidLoad() -> () ()
#7	0x0000000105f4cc99 in -[UIViewController loadViewIfRequired] ()
#8	0x0000000105f4d0cc in -[UIViewController view] ()
#9	0x0000000105e16c51 in -[UIWindow addRootViewControllerViewIfPossible] ()
#10	0x0000000105e173a2 in -[UIWindow _setHidden:forced:] ()
#11	0x0000000105e2acb5 in -[UIWindow makeKeyAndVisible] ()
#12	0x0000000105da3c89 in -[UIApplication _callInitializationDelegatesForMainScene:transitionContext:] ()
#13	0x0000000105da9de9 in -[UIApplication _runWithMainScene:transitionContext:completion:] ()
#14	0x0000000105da6f69 in -[UIApplication workspaceDidEndTransaction:] ()
#15	0x000000010e4b5723 in __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ ()
#16	0x000000010e4b559c in -[FBSSerialQueue _performNext] ()
#17	0x000000010e4b5925 in -[FBSSerialQueue _performNextFromRunLoopSource] ()
#18	0x0000000107f2b311 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ ()
#19	0x0000000107f1059c in __CFRunLoopDoSources0 ()
#20	0x0000000107f0fa86 in __CFRunLoopRun ()
#21	0x0000000107f0f494 in CFRunLoopRunSpecific ()
#22	0x0000000105da57e6 in -[UIApplication _run] ()
#23	0x0000000105dab964 in UIApplicationMain ()
#24	0x0000000104d6d93f in main at /Users/jackson.beachwood/Documents/Development/Source/Projects/Issue Demonstrations/RxTextViewTextColorBug/RxTextViewTextColorBug/AppDelegate.swift:12
#25	0x00000001094f568d in start ()
#26	0x00000001094f568d in start ()
`

**Comment by kzaher** [comment]
Hi @jbeachwood ,

maybe I've expressed myself wrong. What I meant to say:

I've investigated this, but I'm not sure it's possible to solve this problem.

Yes, I have managed to repro it. 

I've pasted the implementation code snippet because it contains explanations why the decision to use `.rx.didProcessEditingRangeChangeInLength` was made, and it also contains explanation why we need to observe all changes => color is one of the triggers for `editedAttributes` change.
